addons/mail/static/src/models/messaging/messaging.js

```diff
@@ -36,79 +87,40 @@ function factory(dependencies) {
+         * Opens the most appropriate view that is a profile for provided id and
+         * model.
          *
          * @param {Object} param0
          * @param {integer} param0.id
          * @param {string} param0.model
-         * FIXME needs to be tested and maybe refactored (see task-2244279)
          */
-        async redirect({ id, model }) {
-            if (model === 'mail.channel') {
-                const channel = this.env.models['mail.thread'].find(thread =>
-                    thread.id === id &&
-                    thread.model === 'mail.channel'
-                );
-                if (!channel || !channel.isPinned) {
-                    this.env.models['mail.thread'].joinChannel(id, { autoselect: true });
-                    return;
-                }
-                channel.open();
-            } else if (model === 'res.partner') {
-                if (id === this.currentPartner.id) {
-                    this.openDocument({
-                        model: 'res.partner',
-                        id,
-                    });
-                    return;
-                }
+        async openProfile({ id, model }) {
+            if (model === 'res.partner') {
                 const partner = this.env.models['mail.partner'].insert({ id });
-                if (!partner.user) {
-                    await this.async(() => partner.checkIsUser());
-                }
-                if (!partner.user) {
-                    // partner is not a user, open document instead
-                    this.openDocument({
-                        model: 'res.partner',
-                        id: partner.id,
-                    });
-                    return;
-                }
-                const chat = partner.correspondentThreads.find(thread =>
-                    thread.channel_type === 'chat'
-                );
-                if (!chat) {
-                    this.env.models['mail.thread'].createChannel({
-                        autoselect: true,
-                        partnerId: id,
-                        type: 'chat',
-                    });
-                    return;
+                return partner.openProfile();
+            }
+            if (model === 'res.users') {
+                const user = this.env.models['mail.user'].insert({ id });
+                return user.openProfile();
+            }
+            if (model === 'mail.channel') {
+                let channel = this.env.models['mail.thread'].findFromIdentifyingData({ id, model: 'mail.channel' });
+                if (!channel) {
+                    channel = (await this.async(() =>
+                        this.env.models['mail.thread'].performRpcChannelInfo({ ids: [id] })
+                    ))[0];
                 }
-                chat.open();
-            } else {
-                this.openDocument({
-                    model,
-                    id,
-                });
+                return channel.openProfile();
```

Comment:
Channel might still be undefined here if it was somehow deleted, should probably consider a notification?

Reply:
Or just preventive/undefined return like for the mother models

Reply:
*edit: Or just let "open document" crash/handle it like for the other models.